Make aot_autograd explicitly error when double backward

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92348

Mitigates https://github.com/pytorch/pytorch/issues/91469

Changes:
- ~once_differentiable can now be parametrized to print a custom error message~
- instead of once_differentiable, we do the backward inside another custom Function, which makes sure the graph is connected, but also makes sure to error on double backward
- we now explicitly error when doing double backward with torch.compile + aot_autograd instead of being silently incorrect. ~The niceness of the error message can vary depending on whether your grad_outputs are passed, or whether you are doing `.grad()` or `.backward()`.~

Unchanged:
- doing backward inside compiled function is still allowed. It currently causes a graph break and is equivalent to doing backward outside the compiled function. It might be nice to disallow this explicitly as well, but that can be done in a follow up.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

soulitzer committed:
Make aot_autograd explicitly error when double backward

[ghstack-poisoned]

soulitzer committed:
Update on "Make aot_autograd explicitly error when double backward"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -332,6 +332,58 @@ def guard_fail_fn(failure):
+            # See @once_differentiable docs for why there are two different errors
+            x = torch.tensor(1.0, requires_grad=True)
+            if grad_output is None:
+                err = "does not require grad and does not have a grad_fn"
```
This error message isn't sufficient because autograd-related subsystems have different behavior depending on if the outputs are not related to the input -- we should always try to raise an error about torch.compile. Here's an example:

Using a strategy like torch.autograd.functional.hessian to compute hessians (functorch.hessian has a similar strategy but also gets a much different error message on this case), we get zero hessians for a once_differentiable function. This would be silently incorrect -- we don't want to give zero second derivative for a torch.compile'd function in any case, we want to give an error message.
```py
import torch

class Func(torch.autograd.Function):
    def forward(ctx, x):
        ctx.save_for_backward(x)
        return x.sin()

    @torch.autograd.function.once_differentiable
    def backward(ctx, gy):
        x, = ctx.saved_tensors
        return gy * x.cos()

x = torch.tensor(1., requires_grad=True)
ggx = torch.autograd.functional.hessian(Func.apply, x)
```

zou3519 replied:
If we can't find a better solution than `@once_differentiable` then I'm fine with the current solution in this PR as a stopgap, because it is less silently incorrect on fewer cases than before.

soulitzer replied:
I think we can improve on `@once_differentiable` if we specialize for this specific case. Updated now.

soulitzer committed:
Update on "Make aot_autograd explicitly error when double backward"


Mitigates https://github.com/pytorch/pytorch/issues/91469

Changes:
- once_differentiable can now be parametrized to print a custom error message
- we now explicitly error when doing double backward with torch.compile + aot_autograd instead of being silently incorrect. The niceness of the error message can vary depending on whether your grad_outputs are passed, or whether you are doing `.grad()` or `.backward()`.

Unchanged:
- doing backward inside compiled function is still allowed, however, it causes a graph break and is equivalent to doing backward outside the compiled function. It might be nice to disallow this explicitly as well, but that can be done in a follow up.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

soulitzer committed:
Update on "Make aot_autograd explicitly error when double backward"


Mitigates https://github.com/pytorch/pytorch/issues/91469

Changes:
- ~once_differentiable can now be parametrized to print a custom error message~
- instead of once_differentiable, we do the backward inside another custom Function, which makes sure the graph is connected, but also makes sure to error on double backward
- we now explicitly error when doing double backward with torch.compile + aot_autograd instead of being silently incorrect. ~The niceness of the error message can vary depending on whether your grad_outputs are passed, or whether you are doing `.grad()` or `.backward()`.~

Unchanged:
- doing backward inside compiled function is still allowed. It currently causes a graph break and is equivalent to doing backward outside the compiled function. It might be nice to disallow this explicitly as well, but that can be done in a follow up.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

soulitzer committed:
Update on "Make aot_autograd explicitly error when double backward"


Mitigates https://github.com/pytorch/pytorch/issues/91469

Changes:
- ~once_differentiable can now be parametrized to print a custom error message~
- instead of once_differentiable, we do the backward inside another custom Function, which makes sure the graph is connected, but also makes sure to error on double backward
- we now explicitly error when doing double backward with torch.compile + aot_autograd instead of being silently incorrect. ~The niceness of the error message can vary depending on whether your grad_outputs are passed, or whether you are doing `.grad()` or `.backward()`.~

Unchanged:
- doing backward inside compiled function is still allowed. It currently causes a graph break and is equivalent to doing backward outside the compiled function. It might be nice to disallow this explicitly as well, but that can be done in a follow up.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1858,22 +1858,39 @@ def backward(ctx, *flat_args):
-                    CompiledFunction.compiled_bw = aot_config.bw_compiler(
-                        bw_module, all_args
+
+            class CustomDelayedError(torch.autograd.Function):
```
AOTAutogradCompiledBackward ?

soulitzer replied:
Sounds good. Two questions/comments:

1. Should we consider the names of these nodes public API (I think so?)
2. I like having "AOTAutograd" in the node name since it is more informative. It feels bad though that the other custom Function is "CompiledFunction". 

I'm okay with landing it like this, but I'm curious if there's anything to do to improve this situation.

albanD reviewed (approved):
nit naming. Sounds good to me otherwise!

soulitzer committed:
Update on "Make aot_autograd explicitly error when double backward"


Mitigates https://github.com/pytorch/pytorch/issues/91469

Changes:
- ~once_differentiable can now be parametrized to print a custom error message~
- instead of once_differentiable, we do the backward inside another custom Function, which makes sure the graph is connected, but also makes sure to error on double backward
- we now explicitly error when doing double backward with torch.compile + aot_autograd instead of being silently incorrect. ~The niceness of the error message can vary depending on whether your grad_outputs are passed, or whether you are doing `.grad()` or `.backward()`.~

Unchanged:
- doing backward inside compiled function is still allowed. It currently causes a graph break and is equivalent to doing backward outside the compiled function. It might be nice to disallow this explicitly as well, but that can be done in a follow up.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1858,22 +1858,39 @@ def backward(ctx, *flat_args):
-                    CompiledFunction.compiled_bw = aot_config.bw_compiler(
-                        bw_module, all_args
+
+            class CompiledFunctionBackward(torch.autograd.Function):
```
Just a thought - since this code is hot-path, would it make sense only generate the custom function when we see that the inputs to the backward pass require gradients?

soulitzer replied:
Good idea. I also realize that maybe_clear_saved_tensors is referring to the wrong ctx now, so that should fix that also.

soulitzer replied:
Updated here: https://github.com/pytorch/pytorch/pull/92688

eellison commented:
This patch led to a large memory regression in timm models: set cudagraphs=False in torch/_inductor/config.py and then run `python benchmarks/dynamo/timm_models.py  --float32 --training --only=mobilevit_s --performance --inductor_no_cudagraphs` 1.40x memory compression -> 1.19 . I tried patching https://github.com/pytorch/pytorch/pull/92688 but still get the memory regression

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1858,22 +1858,39 @@ def backward(ctx, *flat_args):
+                # https://github.com/pytorch/pytorch/pull/92348/files#r1072962107
+                @staticmethod
+                def forward(ctx, *all_args):
+                    all_args_list = list(all_args)
```
hmmm.... don't we need to clear this list 🤔 (@eellison @soulitzer)

bdhirsh replied:
oh actually maybe not, we're calling `call_func_with_args(all_args_list, steal_args=True)` later

soulitzer replied:
I think you might be right... we need to `del all_args` after `all_args_list = list(all_args)`. `all_args_list` will probably be cleared by `call_func_with_args`, but all_args still holds a reference to the tensor.
